proposals/nullable-reference-types-specification.md

```diff
@@ -0,0 +1,266 @@
+
+A given type can have one of four nullabilities: *Oblivious*, *nonnullable*, *nullable* and *unknown*. *Nonnullable* and *unknown* types may yield warnings if a potential `null` value is assigned to them. Values of *nullable* and *unknown* types may yield warnings when dereferenced without proper null checking.
+
+The kind of type and the nullable annotation context it occurs in determine its nullability:
+
+- A nonnullable value type `S` is always *nonnullable*
+- A nullable value type `S?` is always *nullable*
+- An unannotated reference type `C` in a *disabled* annotation context is *oblivious*
+- An unannotated reference type `C` in an *enabled* annotation context is *nonnullable*
+- A nullable reference type `C?` in a *disabled* annotation context is *nullable* (but a warning is yielded)
+- A nullable reference type `C?` in an *enabled* annotation context is *nullable*
+
+Type parameters additionally take their constraints into account:
+
+- An unconstrained type parameter `T` is *unknown*
+- A type parameter `T` where all constraints are *nullable*, *unknown* or the `class?` constraint is *unknown*
+- A type parameter `T` where at least one constraint is *oblivious* or *nonnullable* or one of the `struct` or `class` constraints is
+    - *oblivious* in a *disabled* annotation context
+    - *nonnullable* in an *enabled* annotation context
+- A nullable type parameter `T?` where at least one of `T`'s constraints is *oblivious* or *nonnullable* or one of the `struct` or `class` constraints, is
+    - *nullable* in a *disabled* annotation context (but a warning is yielded)
+    - *nullable* in an *enabled* annotation context
+
+For a type parameter `T`, `T?` is only allowed if `T` is known to be a value type or known to be a reference type.
+
+
+# Constraints
+
+Nullable reference types can be used as generic constraints. Furthermore `object` is now valid as an explicit constraint. Absence of a constraint is now equivalent to an `object?` constraint (instead of `object`), but (unlike `object` before) `object?` is not prohibited as an explicit constraint.
+
+`class?` is a new constraint denoting "possibly nullable reference type", whereas `class` denotes "nonnullable reference type".
+
+The nullability of a type argument or of a constraint does not impact whether the type satisfies the constraint, except where that is already the case today (nullable value types do not satisfy the `struct` constraint). However, if the type argument does not satisfy the nullability requirements of the constraint, a warning may be given.
+
+
+# Null state and null tracking
+
+Every expression in a given source location has a *null state*, which indicated whether it is believed to potentially evaluate to null. The null state is either "not null" or "maybe null". The null state is used to determine whether a warning should be given about null-unsafe conversions and dereferences.
+
```

Comment:
is there value in having a "definitely null" null-state?  for example:

```if (param != null) { ... } else { $param$... }```

if we know something to be 'definitely null', there could be specific warnings emitted that the user coudl then decide should be treated always as errors.